modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -224,16 +229,75 @@ private void doWriteChunkedResponse(ChannelHandlerContext ctx, Netty4ChunkedHttp
             return;
         }
-        assert currentChunkedWrite.responseBody().isDone();
         final var finishingWrite = currentChunkedWrite;
         currentChunkedWrite = null;
-        writeSequence++;
-        finishingWrite.combiner().finish(finishingWrite.onDone());
+        final var finishingWriteBody = finishingWrite.responseBody();
+        assert finishingWriteBody.isDone();
+        final var endOfResponse = finishingWriteBody.isEndOfResponse();
+        if (endOfResponse) {
+            writeSequence++;
+            finishingWrite.combiner().finish(finishingWrite.onDone());
+        } else {
+            final var channel = finishingWrite.onDone().channel();
+            ActionListener.run(ActionListener.assertOnce(new ActionListener<>() {
+                @Override
+                public void onResponse(ChunkedRestResponseBody continuation) {
+                    channel.writeAndFlush(
+                        new Netty4ChunkedHttpContinuation(writeSequence, continuation, finishingWrite.combiner()),
+                        finishingWrite.onDone() // pass the terminal listener/promise along the line
+                    );
+                    checkShutdown();
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.error(
+                        Strings.format("failed to get continuation of HTTP response body for [%s], closing connection", channel),
+                        e
+                    );
+                    channel.close().addListener(ignored -> {
+                        finishingWrite.combiner().add(channel.newFailedFuture(e));
+                        finishingWrite.combiner().finish(finishingWrite.onDone());
+                    });
+                    checkShutdown();
+                }
+
+                private void checkShutdown() {
+                    if (channel.eventLoop().isShuttingDown()) {
+                        // The event loop is shutting down, and https://github.com/netty/netty/issues/8007 means that we cannot know if the
+                        // preceding activity made it onto its queue before shutdown or whether it will just vanish without a trace, so
+                        // to avoid a leak we must double-check that the final listener is completed once the event loop is terminated.
+                        // Note that the final listener came from Netty4Utils#safeWriteAndFlush so its executor is an ImmediateEventExecutor
+                        // which means this completion is not subject to the same issue, it still works even if the event loop has already
+                        // terminated.
+                        channel.eventLoop()
+                            .terminationFuture()
+                            .addListener(ignored -> finishingWrite.onDone().tryFailure(new ClosedChannelException()));
+                    }
+                }
+
+            }), finishingWriteBody::getContinuation);
```

Comment:
Given that `getContinuation` is expected to be async and take a while, I wonder if we should add some sort of prefetching of these to minimize the stop-go sending here? Clearly something we can do in a follow-up, this is fine as is for now.

Reply:
Yes, but I was expecting to push that responsibility onto the implementation(s) since they will know much more about the costs & benefits of doing so. For instance, we could use a queue to allow parts to arrive asynchronously, and only return `true` from `isDone` if the queue is empty.

Reply:
Also some implementations will be able to handle certain reorderings of parts (e.g. entries in a `.zip` file) but there's no reasonable way to expose that kind of subtlety here IMO.